Source/WebCore/dom/Document.cpp

```diff
@@ -3432,23 +3432,19 @@ AXObjectCache* Document::axObjectCache() const
     // document.  This is because we need to be able to get from any WebCoreAXObject
     // to any other WebCoreAXObject on the same page.  Using a single cache allows
     // lookups across nested webareas (i.e. multiple documents).
-    RefPtr mainFrameDocument = this->mainFrameDocument();
-
-    if (!mainFrameDocument) {
-        LOG_ONCE(SiteIsolation, "Unable to reach the main frame documents AXObjectCache ");
-        return nullptr;
-    }
+    RefPtr rootFrame = m_frame ? &m_frame->rootFrame() : nullptr;
+    RefPtr rootFrameDocument = rootFrame ? rootFrame->document() : nullptr;
 
     // If the document has already been detached, do not make a new axObjectCache.
-    if (!mainFrameDocument->hasLivingRenderTree())
+    if (!rootFrameDocument || !rootFrameDocument->hasLivingRenderTree())
         return nullptr;
 
-    ASSERT(mainFrameDocument.get() == this || !m_axObjectCache);
-    if (!mainFrameDocument->m_axObjectCache) {
-        mainFrameDocument->m_axObjectCache = makeUnique<AXObjectCache>(*mainFrameDocument);
+    ASSERT(rootFrameDocument.get() == this || !m_axObjectCache);
+    if (!rootFrameDocument->m_axObjectCache) {
+        rootFrameDocument->m_axObjectCache = makeUnique<AXObjectCache>(*rootFrameDocument);
```

Comment:
I think this is an incorrect change in behavior.

The purpose of always storing the `AXObjectCache `on the main frame document is that so - for any given frame tree - there's exactly one `AXObjectCache`.

My reading of `rootFrame` code is that there can be multiple root frames in a given frame tree.
e.g. One LocalFrame in a site isolated process might end up with itself as the `rootFrame`, and a sibling `LocalFrame` could end up with **itself** as its `rootFrame`

I believe the proper fix for the `AXObjectCache` storage is that it lives on `Page` and is cleared when `Page::didChangeMainDocument()` is called.